ipld/get_namespaced_shares.go

```diff
@@ -0,0 +1,226 @@
+		lowest = atomic.LoadInt64(&b.lowest)
+	}
+	// we always run both checks because element can be both the lower and higher bound
+	// for example, if there is only one share in the namespace
+	highest := atomic.LoadInt64(&b.highest)
+	for index > highest && !atomic.CompareAndSwapInt64(&b.highest, highest, index) {
+		highest = atomic.LoadInt64(&b.highest)
+	}
+}
+
+// getLeavesByNamespace returns all the leaves from the given root with the given namespace.ID.
+// If no shares are found, it returns both data and error as nil.
+// A non-nil error means that only partial data is returned, because at least one share retrieval failed
+// The following implementation is based on `GetShares`.
+func getLeavesByNamespace(
+	ctx context.Context,
+	bGetter blockservice.BlockGetter,
+	root cid.Cid,
+	nID namespace.ID,
+	maxShares int,
+) ([]ipld.Node, error) {
+	ctx, span := tracer.Start(ctx, "get-leaves-by-namespace")
+	defer span.End()
+
+	span.SetAttributes(
+		attribute.String("namespace", nID.String()),
+		attribute.String("root", root.String()),
+	)
+
+	err := SanityCheckNID(nID)
+	if err != nil {
+		return nil, err
+	}
+
+	// we don't know where in the tree the leaves in the namespace are,
+	// so we keep track of the bounds to return the correct slice
+	// maxShares acts as a sentinel to know if we find any leaves
+	bounds := fetchedBounds{int64(maxShares), 0}
+
+	// buffer the jobs to avoid blocking, we only need as many
+	// queued as the number of shares in the second-to-last layer
+	jobs := make(chan *job, (maxShares+1)/2)
+	jobs <- &job{id: root}
+
+	var wg wrappedWaitGroup
+	wg.jobs = jobs
+	wg.Add(1)
+
+	var singleErr sync.Once
+	var retrievalErr error
+
+	// we overallocate space for leaves since we do not know how many we will find
+	// on the level above, the length of the Row is passed in as maxShares
+	leaves := make([]ipld.Node, maxShares)
+
+	// stillWorking will be set to false when the waitgroup counter reaches 0, closing the jobs channel
+	for {
+		select {
+		case j, ok := <-jobs:
+			if !ok {
+				// we didn't find any leaves, return nil
+				if bounds.lowest == int64(maxShares) {
+					return nil, retrievalErr
+				}
+
+				return leaves[bounds.lowest : bounds.highest+1], retrievalErr
+			}
+			namespacePool.Submit(func() {
+				ctx, span := tracer.Start(ctx, "process-job")
+				defer span.End()
+				defer wg.Done()
+
+				// if an error is likely to be returned or not depends on
+				// the underlying impl of the blockservice, currently it is not a realistic probability
+				nd, err := plugin.GetNode(ctx, bGetter, j.id)
+				if err != nil {
+					singleErr.Do(func() {
+						log.Errorw("getSharesByNamespace: could not retrieve node", "nID", nID, "pos", j.pos, "err", err)
+						retrievalErr = err
+					})
+					span.RecordError(err, trace.WithAttributes(
+						attribute.Int("pos", j.pos),
+					))
+					// we still need to update the bounds
+					bounds.Update(int64(j.pos))
```

Comment:
Hmm, can this corrupt the resulting boundary if this fails for any inner node?

Reply:
Can you give some more clarification as to what you mean by corruption? If it fails for an inner node, the array may possibly not return as many children as it would (it is not extended outwards), but it wouldn't corrupt anything no?

The reason the update is even needed here is for an edge case I found during testing: If the **last** node is nil, then the bounds are not updated and it is not kept in the result array.